src/background/background-process.js

```diff
@@ -228,27 +234,64 @@ export default class BackgroundProcess {
   }
 
   loadContentData (tabScript) {
-    let polyfillScript = this.loadPolyfill(tabScript.tabID)
-    let contentScript = this.loadContentScript(tabScript.tabID)
+    let polyfillScript = this.loadPolyfill(tabScript.tabObj.tabId)
+    let contentScript = this.loadContentScript(tabScript.tabObj.tabId)
     let contentCSS = []
     for (let fileName of this.settings.contentCSSFileNames) {
-      contentCSS.push(this.loadContentCSS(tabScript.tabID, fileName))
+      contentCSS.push(this.loadContentCSS(tabScript.tabObj.tabId, fileName))
     }
     return Promise.all([polyfillScript, contentScript, ...contentCSS])
   }
 
   stateMessageHandler (message, sender) {
     let contentState = TabScript.readObject(message.body)
+    contentState.updateTabObject(sender.tab.id, sender.tab.windowId)
+
     this.updateTabState(contentState.tabID, contentState)
-    console.log(this.tabs.get(contentState.tabID))
   }
 
   tabActivationListener (info) {
-    this.tab = info.tabId
-    let tab = this.tabs.has(info.tabId) ? this.tabs.get(info.tabId) : undefined
+    let tmpUniqueTabId = Tab.createUniqueId(info.tabId, info.windowId)
+    this.tab = tmpUniqueTabId
+    let tab = this.tabs.has(tmpUniqueTabId) ? this.tabs.get(tmpUniqueTabId) : undefined
+
     this.setMenuForTab(tab)
   }
 
+  tabDetachedListener (tabId, detachInfo) {
+    let tmpUniqueTabObj = Tab.createUniqueId(tabId, detachInfo.oldWindowId)
+    if (this.tabs.has(tmpUniqueTabObj)) {
+      this.tabs.get(tmpUniqueTabObj).deattach()
+    }
+  }
+
+  tabAttachedListener (tabId, attachInfo) {
```

Comment:
can you add some comments to explain what this method is doing exactly? I'm not sure I understand the code. I think what you are doing is that if the tab exists in our state, either in a detached state or a non-detached state, we reinitialize it with the new tab id.  Is that correct?

Reply:
There are two 2 workflows with attached and deattached:

1) Tab is a part of win1 - we are deattaching it to a separate window itself, this way two events creates - 
     first deattached (tabDetachedListener) and i marked it with status property
     second attached (tabAttachedListener) and I find deattached tab in array and update windowId property (first forEach)

2) Tab is in a separate window - and I attached it to the existed window.
    In this case only attached event created  (tabAttachedListener) so I use the second forEach and search for the tab with the same tabId and updating it.